src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMapsFree/src/ElectricityMapsFreeDataSource.cs

```diff
@@ -86,15 +86,27 @@ public async Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(Location l
         }
         
         List<EmissionsData> EmissionsDataList = new List<EmissionsData>();
+        var emissionDateTime = gridEmissionData.Data.Datetime;
+        var shouldReturn = true;
+        if (emissionDateTime != null && periodStartTime < periodEndTime)
```

Comment:
Agree, so is it better to throw exception (e.g. [ArgumentOutOfRangeException](https://learn.microsoft.com/ja-jp/dotnet/api/system.argumentoutofrangeexception?view=net-7.0) ) instead of returning empty list in that case? I think it is an illegal case if `GetCarbonIntensityAsync()` receives future datetime, so I think it is appropriate to throw an exception.

Reply:
I think other datasources (`WattTimeDataSource`, `ElectricityMapsDataSource`) have same problems. If it is proper to throw exception, it is better to handle another issue/PR to fix all of datasources.

Reply:
Below is for WattTime (I requested 1 hour into the future, on the emissions/bylocation endpoint), so it is returning 204 (which is to be expected?) 
<img width="692" alt="Screenshot 2023-09-19 at 07 53 39" src="https://github.com/Green-Software-Foundation/carbon-aware-sdk/assets/26546660/087bea7c-28c4-4e9a-a8e3-294304614bb9">

The way this is implemented in the WattTimeDataSource.cs, is that they use a helper class `IntervalHelper` which is just a friendly wrapper around a clamping filter:
https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/dev/src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs#L72

https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/dev/src/CarbonAware/src/IntervalHelper.cs

Its simple, but maybe we can reuse it here?

Reply:
So as discussed in #394 - for now align to the WattTIme DataSource (so return empty array on invalid date range/future date range) using IntervalHelper. Create a new issue to consider throwing exception on requesting future data on the historical endpoints

Reply:
I pushed new commit to return empty list when ElectricityMapsFree.GetCarbonIntensityAsync() receives request for forecast. AFAICS `IntervalHelper` expects array of `EmissionsData` which includes emission datetime, however EMFree seems to be possible to lack it.

https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/cfd76fe2e0e5b1f1d32bd6a8faf4c7e7a1c245a0/src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMapsFree/src/ElectricityMapsFreeDataSource.cs#L90-L96

And also `EmissionsHandler` would set same value (`UtcNow`) when `startTime` and `endTime` is `null`.

https://github.com/Green-Software-Foundation/carbon-aware-sdk/blob/cfd76fe2e0e5b1f1d32bd6a8faf4c7e7a1c245a0/src/GSF.CarbonAware/src/Handlers/EmissionsHandler.cs#L50-L51

Hence I fixed to check whether `periodStartTime` is less than current time, and returns empty list if so.